app/Metadata/Extractor.php

```diff
@@ -477,4 +486,22 @@ public function extract(string $fullPath, string $kind): array
 
 		return $metadata;
 	}
+
+	/**
+	 * https://github.com/LycheeOrg/Lychee/issues/1200.
+	 *
+	 * Ensures the provided DateTime value is functional
+	 *
+	 * @param bool|\DateTime $d
+	 *
+	 * @return bool
+	 */
+	protected function date_is_valid(bool|\DateTime $d): bool
+	{
+		if (false === $d) {
+			return false;
+		}
+
+		return self::MINIMUM_VALID_DATE_UTC <= intval($d->format('U'));
```

Comment:
That is the condition I feared above.

While I understand that `0000-00-00 00:00:00` is not a valid date, because there is no zero month and day. I wonder why a valid `\DateTime` should print such a date when converted to a string. When I first read the issue #1200, I thought that maybe the EXIF extractor returned a ill-formatted string `0000-00-00 00:00:00` if this was the EXIF data. But here we already have a valid `\DateTime` instance.

Also please don't use the conversion chain `\DateTime` --> `string` --> `int`. This is just unnecessary and bad. It will even break on 32bit systems after 2038. If we really needed to enforce a lower bound on the timestamp, then use the appropriate methods of the `\DateTime` interface for comparison. You may even use the normal mathematical operators `<`, `<=`, `>=` and `>` for comparison between two `\DateTime` objects, because these operators are overloaded for `\DateTime`. (Yes, PHP allows operator overloading just like C++).

While I believe you that this might solve the bug for you, I am still not convinced that this is the least invasive solution.

Reply:
So technically speaking exif values like that are just ascii fields.  Without some kind of external control, the value can be whatever you want.   The Nikon in question, and I assume most / all Nikons of the era and possibly other cameras, have an invalid default value.

[\DateTime will happily consume a 0000:00:00 value](https://3v4l.org/SU8lg), and _technically_ it is responsible for further mutating the value to the one seen in DB errors.

| Also please don't use the conversion chain \DateTime --> string --> int. This is just unnecessary and bad.

I beg to differ.  In my opinion, an integer representing a Unix timestamp is the absolute easiest and least error-prone way to programmatically represent time.  However, this is not my project and I don't really feel like arguing on this point.

Additionally, post 2038 32bit systems will have a lot more to worry about than an exif field sanity check :)

I have a test setup for pgsql and mariadb locally now, so I will make some updates tonight or tomorrow.

Reply:
> Additionally, post 2038 32bit systems will have a lot more to worry about than an exif field sanity check :)

This doesn't mean that we need to contribute to that problem.

>> Also please don't use the conversion chain \DateTime --> string --> int. This is just unnecessary and bad.

> I beg to differ. In my opinion, an integer [...] is the absolute easiest and least error-prone way to programmatically represent time.

This doesn't impu that you have to convert it to a string first.

> \DateTime will happily consume a 0000:00:00 value, and technically it is responsible for further mutating the value to the one seen in DB errors.

As you can see from your own example, DateTime accepts that input, but already normalizes it to the nearest valid value and stores that value internally.

That is exactly what I assumed in the first place. If we already have a DateTime object, the test is too late, because anything which can be represented by DateTime will already be a valid date. So the problem must be fixed at some complete different place, i.e. inside the exif extractor _before_ the DateTime object is even constructed and we still have access to the original ASCII string from the uninterpreted, raw EXIF data.